### DISCUSSION TITLE
Next/dynamic doesn't appear to support compound components

### DISCUSSION THREAD

**Discussion by pujiragil** [vercel/next.js#51297] [Help] [answered]
### Summary

I want to use dynamic import on the ProductCard component, but the issue is that I'm using the compound component pattern in that component. When I tried implementing the ProductCard component with dynamic import and React.lazy, I encountered an error. However, when I use regular import for that component, everything works fine.

```javascript
import dynamic from "next/dynamic";
import { dresses } from "../../utils/dress";
import { DetailButton } from "../Global";
import { CatalogSidebar } from "./catalogSidebar";

const ProductCard = dynamic(() => import("../Card/productCard"), {
  loading: () => <p>Loading masbro...</p>,
});

export default function Catalog() {
  return (
    <div className="mx-3.5 mb-3.5 gap-x-8 font-oswald lg:mb-8 lg:flex">
      <h3 className="mb-4 text-lg md:text-xl lg:hidden">Shop Some Wear:</h3>

      <div className="hidden w-3/12 lg:block">
        <CatalogSidebar />
      </div>

      <div className="w-full lg:pb-[18px]">
        <div className="mb-[18px] grid grid-cols-2 gap-3.5 sm:grid-cols-3 md:grid-cols-4 lg:mb-12 lg:gap-6">
          {dresses.map((dress) => (
            <ProductCard
              key={dress.id}
              data={dress}
              image={<ProductCard.Image />}
              info={<ProductCard.Info />}
              price={<ProductCard.Price />}
            />
          ))}
        </div>
        <DetailButton />
      </div>
    </div>
  );
}
```

here's the error message

`
Unhandled Runtime Error

Error: Element type is invalid: expected a string (for built-in components) or a class/function (for composite components) but got: undefined. You likely forgot to export your component from the file it's defined in, or you might have mixed up default and named imports.

Check the render method of `Catalog`.
`

### Additional information

```js
Unhandled Runtime Error

Error: Element type is invalid: expected a string (for built-in components) or a class/function (for composite components) but got: undefined. You likely forgot to export your component from the file it's defined in, or you might have mixed up default and named imports.

Check the render method of `Catalog`.
```

### Example

https://stackblitz.com/edit/stackblitz-starters-qv8kfg?file=pages%2Findex.js

**Comment by icyJoseph** [comment]
Hey, yeah this is a bit convoluted, but the dynamic pattern wraps your component with a "Loadable" which controls the loading/resolve/reject of getting that module. It then takes the default export and renders that. So it behaves, as a HOC.

However, that "Loadable" is never your ProductCards component.

In other words, say you use `import { memo } from "react"`, like this:

```js
const ProductCard = memo(OriginalProductCard)
```

Would you expect `ProductCard.Image` to work? No right? Because the `memo` HoC has wrapped it. You are not working with `OriginalProductCard` anymore. In this case though, ONLY FOR MEMO, you could use `ProductCard.type.Image`, because the component is in the `type` property, and that's where you can make a reference to the compound component.

The "Loadable" exposes, AFAIK, nothing out of its inner component.

**Reply by pujiragil** [reply]
So, is there no solution for using compound components? Are there any alternative approaches?

**Reply by icyJoseph** [reply]
Let me try a few things

**Reply by icyJoseph** [reply, marked as answer]
Ah nope, I though the `render` function was available, but it is not. So we need to hack this, it is a combination of two approaches:

- https://stackblitz.com/edit/stackblitz-starters-2edjkw?file=components%2FproductCard.js,components%2Fcatalog.js

I moved ProductCard after all supporting components, and defined fallbacks. I also take them as components `() => <div />`, not elements `<div />`:

```jsx
const ProductCard = ({ image, info, price, color, data }) => {
  const ImageChild = image || ProductImage;
  const Info = info || ProductInfo;
  const Price = price || ProductPrice;
  const Color = color || ProductColor;

  return (
    <ProductCardContext.Provider value={data}>
      <div className="space-y-2">
        <ImageChild />
        <div className="space-y-2">
          <Info />
          <Price />
          <Color />
        </div>
      </div>
    </ProductCardContext.Provider>
  );
};

ProductCard.Image = ProductImage;
ProductCard.Info = ProductInfo;
ProductCard.Price = ProductPrice;
ProductCard.Color = ProductColor;
export default ProductCard;
```

And then from the dynamic import side:

```jsx
const ProductCard = dynamic(
  async () => {
    const mod = await import('./productCard');

    ProductCard.Image = mod.default.Image;
    ProductCard.Info = mod.default.Info;
    ProductCard.Price = mod.default.Price;

    return mod.default;
  },
  {
    loading: () => <p>Loading masbro...</p>,
  }
);

export default function Catalog() {
  return (
      <div className="w-full lg:pb-[18px]">
        <div className="mb-[18px] grid grid-cols-2 gap-3.5 sm:grid-cols-3 md:grid-cols-4 lg:mb-12 lg:gap-6">
          {dresses.map((dress) => (
            <ProductCard
              key={dress.id}
              data={dress}
              image={ProductCard.Image}
              info={ProductCard.Info}
              price={ProductCard.Price}
            />
          ))}
        </div>
      </div>
  );
}
```

As said, it is a hack... ugly and a bit inefficient perhaps, but it works.

**Reply by pujiragil** [reply]
I really appreciate your help and the code snippet you shared. It may be a hack, but it works perfectly for my needs. Thanks again!
